skipchain/skipchain.go

```diff
@@ -43,15 +43,63 @@ func init() {
 // Service handles adding new SkipBlocks
 type Service struct {
 	*onet.ServiceProcessor
-	db             *SkipBlockDB
-	propagate      messaging.PropagationFunc
-	verifiers      map[VerifierID]SkipBlockVerifier
-	newBlocksMutex sync.Mutex
-	newBlocks      map[string]bool
-	storageMutex   sync.Mutex
-	Storage        *Storage
-	bftTimeout     time.Duration
-	propTimeout    time.Duration
+	db           *SkipBlockDB
+	propagate    messaging.PropagationFunc
+	verifiers    map[VerifierID]SkipBlockVerifier
+	storageMutex sync.Mutex
+	Storage      *Storage
+	bftTimeout   time.Duration
+	propTimeout  time.Duration
+	chains       chainLocker
+}
+
+type chainLocker struct {
+	sync.Mutex
+	// the key type is string because []byte is not allowed
+	// in Go maps as keys.
+	chains map[string]*sync.Mutex
+	// a count of how many locks are currently held
+	locks int
+}
+
+var errTimeout = errors.New("timeout waiting to lock chain")
+
+func (cl *chainLocker) lock(chain SkipBlockID) {
+	cl.Lock()
+	// Lazy initializtion.
+	if cl.chains == nil {
+		cl.chains = make(map[string]*sync.Mutex)
+	}
+
+	if l, ok := cl.chains[string(chain)]; ok {
+		cl.locks++
+		cl.Unlock()
+		l.Lock()
+		return
+	}
+
+	l := new(sync.Mutex)
+	l.Lock()
+	cl.chains[string(chain)] = l
```

Comment:
When I first wrote this code I thought about it and decided that there is no non-racy way to remove these, so I convinced myself they would have to stay allocated. But rethinking it, I have an idea I'll try.

Reply:
Just wrote this to explain why not:
```
	// It is not possible to delete the entry from the map in a non-racy way.
	// Consider 3 goroutines. #1 has the lock and is here unlocking. #2 is
	// sleeping on the lock. #3 has just received a block for this chain, but has
	// not tried to lock it. When #1 unlocks, #2 wakes up and runs, locking the
	// same mutex that #1 just unlocked. If #1 deletes the mutex from the map,
	// later #2 will unlock it and could just ignore the fact that it is no longer
	// in the map. But if #3 tries to lock while #2 has it locked, but after #1 has
	// removed it from the map, #3 makes a new mutex, and gets the lock. Now #2 and #3
	// are both holding different locks that they each think is the lock for chainId.
	// Not ok.
```

Reply:
I'm sorry to drop it like this but this is really bad (and not your fault too !). Settling for a unbounded growing-only map because of a concurrency design issue in the code is bad. Please let us not forget that it's here and demanding to be solved later on !!

Reply:
One way that may fix this is to start a go-routine and periodically check if the number of locks is 0. If it is, then no requests are waiting to be completed and we can safely empty the map. OTOH,I  don't know how the tests are going to work because they fail on leaky go-routines.

Reply:
Something else to remember: These locks are only taken after we have confirmed that this is a skipchain-id that we know, and there is at most one lock per skipchain. So this map cannot be filled by an attacker, and in our expected use case, where # of skipchains << # of blocks, this is an acceptable cost.

There would be some kind of actor-based way to implement this, where the lock request was sent in to a central lock manager thing via a request channel. But it's just not worth the effort.